File: Serie1PC/Serie1PC/RendezvousChannel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.Design;
using System.Threading;



namespace Serie1PC
{
    public class RendezvousChannel<S,R>
    {
        
        public class Token
        {
            public S service;
            public R response;

            public Token(S service)
            {
                this.service = service;
                this.response = default(R);
            }

      

            public Token()
            {
                service = default(S);
                response = default(R);
            }
        }
        private class Ticket
        {
            public Token token;
            internal Status processing;
            public object _condition;

            public Ticket()
            {
                token = new Token();
                processing = Status.Open;
            }
        }

        private readonly Object _lock = new Object();
        private readonly LinkedList<Ticket> _services;
        private readonly LinkedList<Ticket> _processingRequests;

            public enum Status
            {
                Open,
                Accepting,
                InProcess,
                Concluded
            };
        public RendezvousChannel()
        {
            _services=new LinkedList<Ticket>();
            _processingRequests = new LinkedList<Ticket>();
        }

        public bool Request(S service,int timeout, out R response)
        {
           lock(_lock){
                    Ticket myTicket;
                    if (CheckForAcceptingTokens(out myTicket, service))
                    {
                     SyncUtils.Notify(_lock,myTicket._condition);
                    }
                    else
                    {
                          myTicket.token = new Token(service);
                        myTicket._condition = new Object();
                        _services.AddLast(myTicket);
                    }
                  
                    while (true)
                    {
                        SyncUtils.Wait(_lock,myTicket._condition);
                        if (myTicket.processing == Status.Concluded)
                        {
                            response = myTicket.token.response;
                            return true;
                        }
                    }
            }
            
        }
        private bool CheckForAcceptingTokens( out Ticket myTicket, S service)
        {
            foreach (var ticket in _services)
            {
                if (ticket.processing == Status.Accepting)
                {
                    ticket.token.service = service;
                    myTicket = ticket;
                    ticket.processing=Status.InProcess;
                    _services.Remove(ticket);
                    _processingRequests.AddLast(ticket);
                    return true;
                }
            }
            myTicket=new Ticket();
            return false;
        }

        public  object Accept(int timeout, out S service)
            {
            lock (_lock)
            {
                    var myTicket = new Ticket();
                    Token token = new Token();
                    if (_services.Count != 0)
                    {
                        RemoveFirstRequestedToken(ref myTicket);
                        service = myTicket.token.service;
                        token = myTicket.token;
                        return token;
                    }

                    myTicket._condition = new Object();
                    myTicket.processing = Status.Accepting;
                    _services.AddLast(myTicket);
                    

                    while (true)
                    {
                        SyncUtils.Wait(_lock, myTicket._condition);
                        if (myTicket.processing == Status.InProcess)
                        {
                            service = myTicket.token.service;
                            return myTicket.token;
                        }
                    }
          }
    }
    public void Reply(object rendezVousToken, R Response)
        {
        lock (_lock)
        {
                var myToken = (Token) rendezVousToken;
                FindTokenInProcessingRequests(ref myToken, Response);       
        }
        
        }
        private void RemoveFirstRequestedToken(ref Ticket myTicket)
        {
            foreach (var ticket in _services)
            {
                if (ticket.processing == Status.Open)
                {
                    myTicket = ticket;
                    ticket.processing = Status.InProcess;
                    _services.Remove(ticket);
                    _processingRequests.AddLast(ticket);
                    return;
                }

            }
            
        }
        private void FindTokenInProcessingRequests(ref Token myToken, R response)
        {

            foreach (var ticket in _processingRequests)
            {
                if (ticket.token.Equals(myToken))
                {
                    myToken.response = response;
                    ticket.processing = Status.Concluded;
                    _processingRequests.Remove(ticket);
                    SyncUtils.Notify(_lock,ticket._condition);           
                    return;
                }
            }
        }
    }
}

File: 1/src/main/java/concurrency/RendezvousChannel.java
package concurrency;

import java.util.LinkedList;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class RendezvousChannel<S,R> {	
	public class Token
	{
		public S service;
		public R response;

		public Token(S service)
		{
			this.service = service;
			this.response = null;
		}



		public Token()
		{
			service = null;
			response = null;
		}
	}
	private class Ticket
	{
		public Token token;
		private Status processing;
		public Condition _condition; // Lock local

		public Ticket()
		{
			token = new Token();
			processing = Status.Open;
		}
	}

	private final ReentrantLock _lock; 
	private final LinkedList<Ticket> _services;
	private final LinkedList<Ticket> _processingRequests;

	public enum Status
	{
		Open,
		Accepting,
		InProcess,
		Concluded
	};

	public RendezvousChannel()
	{
		_lock=new ReentrantLock();
		_services=new LinkedList<Ticket>();
		_processingRequests = new LinkedList<Ticket>();
	}

	public R Request(S service,int timeout) throws InterruptedException
	{
		try
		{
			_lock.lock();
			// lock
			Ticket myTicket;
			if ((myTicket=CheckForAcceptingTokens(service))!=null)
				myTicket._condition.signal();
			else
			{
				myTicket=new Ticket();
				myTicket.token.service=service;
				myTicket._condition = _lock.newCondition();
				_services.addLast(myTicket);
			}

			while (true)
			{
				myTicket._condition.await();
				if (myTicket.processing == Status.Concluded)
				{
					R response = myTicket.token.response;
					return response;
				}
			}
		}
		finally{
			_lock.unlock();
		}

	}
	private Ticket CheckForAcceptingTokens(S service)
	{
		for(Ticket ticket : _services)
		{
			if (ticket.processing == Status.Accepting)
			{
				ticket.token.service = service;
				ticket.processing=Status.InProcess;
				_services.remove(ticket);
				_processingRequests.addLast(ticket);
				return ticket;
			}
		}
		return null;
	}

	public  Object Accept(int timeout) throws InterruptedException
	{
		try{
			_lock.lock();
			
			if (!_services.isEmpty())
				return RemoveFirstRequestedToken();
			
			Ticket myTicket = new Ticket();
			myTicket._condition=_lock.newCondition();
			myTicket.processing = Status.Accepting;
			_services.addLast(myTicket);
			while (true)
			{
				myTicket._condition.await();
				if (myTicket.processing == Status.InProcess)
					return myTicket.token;

			}
		}
		finally{
			_lock.unlock();
		}
	}
	public void Reply(Object rendezVousToken, R Response)
	{
		try{
			_lock.lock();
			Token myToken = (Token) rendezVousToken;
			FindTokenInProcessingRequests(myToken, Response);       
		}
		finally{
			_lock.unlock();
		}

	}
	private Token RemoveFirstRequestedToken()
	{
		for (Ticket ticket :_services)
		{
			if (ticket.processing == Status.Open)

				ticket.processing = Status.InProcess;
			_services.remove(ticket);
			_processingRequests.addLast(ticket);
			return ticket.token;
		}
		return null;
	}

	private void FindTokenInProcessingRequests(final Token myToken, R response)
	{

		for(Ticket ticket :_processingRequests)
		{
			if (ticket.token.equals(myToken))
			{
				myToken.response = response;
				ticket.processing = Status.Concluded;
				_processingRequests.remove(ticket);
				ticket._condition.signal();           
				return;
			}
		}
	}
}

